pkg/controller/migration/convert/network.go

```diff
@@ -98,22 +116,167 @@ func handleCalicoCNI(c *components, install *Installation) error {
 		install.Spec.CalicoNetwork.HostPorts = &hp
 	}
 
-	// CNI IPAM pools
-	if len(c.calicoCNIConfig.IPAM.IPv4Pools) != 0 {
-		return ErrIncompatibleCluster{"specifying ipv4_pools in cni config is not supported."}
+	if c.cniConfigName != "k8s-pod-network" {
+		return ErrIncompatibleCluster{fmt.Sprintf("%s, only 'k8s-pod-network' is supported as CNI name, found %s", errCtx, c.cniConfigName)}
 	}
 
 	// Other CNI features
 	if c.calicoCNIConfig.FeatureControl.FloatingIPs {
-		return ErrIncompatibleCluster{"floating IPs not supported"}
+		return ErrIncompatibleCluster{errCtx + ", floating IPs not supported"}
 	}
 	if c.calicoCNIConfig.FeatureControl.IPAddrsNoIpam {
-		return ErrIncompatibleCluster{"IpAddrsNoIpam not supported"}
+		return ErrIncompatibleCluster{errCtx + ", IpAddrsNoIpam not supported"}
+	}
+	if c.calicoCNIConfig.ContainerSettings.AllowIPForwarding {
+		return ErrIncompatibleCluster{errCtx + ", AllowIPForwarding not supported"}
+	}
+
+	return nil
+}
+
+func getNetworkingBackend(node CheckedDaemonSet, client client.Client) (string, error) {
+	netBackend, err := node.getEnv(ctx, client, containerCalicoNode, "CALICO_NETWORKING_BACKEND")
+	if err != nil {
+		return "", err
+	}
+
+	switch {
+	case netBackend == nil:
+		return "bird", nil
+	case *netBackend == "":
+		return "bird", nil
+	case *netBackend == "bird":
+		return "bird", nil
+	case *netBackend == "vxlan":
+		return "vxlan", nil
+	case *netBackend == "none":
+		return "none", nil
+	default:
+		return "", fmt.Errorf("CALICO_NETWORKING_BACKEND %s is not valid", *netBackend)
+	}
+}
+
+// subhandleCalicoIPAM checks all fields in the Calico IPAM configuration,
+// if any fields have unexpected values an error message will be returned.
+// The function tries to collect all the errors and report one message.
+// If there are no errors and the config can be added to the passed in 'install'
+// then nil is returned.
+func subhandleCalicoIPAM(node CheckedDaemonSet, client client.Client, cnicfg calicocni.NetConf, install *Installation) error {
```

Comment:
I think this looks much cleaner by not passing in the entire components struct. However:
- I don't think the variable marking functionality of the `CheckedDaemonset` will work if you're passing this by value in to this function. It needs a pointer so that subsequent calls to `getEnv` will update the correct map.
- However, I notice that the only place `node` and `client` are used are during the call to `getNetworkingBackend`. Perhaps you just want call `getNetworkingBacked` in the caller, and then pass in the networking backend only?

Reply:
CheckedDaemonset holds the ignored information in a map which is a reference so when it is copied the copy can still update the ignore information.
I've pulled out the getNetworkingBackend so only the backend value is passed into the subhandle functions.